[test_math.py] | Sanity tests | ttnn-unit-tests / ttnn eltwise group 3 [wh_n300_civ2]

## CI Failure Report

<!-- AUTO-TRIAGE-SUMMARY-START -->
**Summary:** [CI] ttnn eltwise test_cbrt[w=128-h=64] fails with ULP accuracy regression on wh_n300_civ2
<!-- AUTO-TRIAGE-SUMMARY-END -->

**Workflow:** Sanity tests
**Job:** ttnn-unit-tests / ttnn eltwise group 3 [wh_n300_civ2]

### Regression timeline

- **Last passing:** commit `043d369d3446` — 2026-06-29T19:17:29Z — [run](https://github.com/tenstorrent/tt-metal/actions/runs/28396665515)
- **First failing:** commit `7f36d0321365` — 2026-06-29T21:10:54Z — [job](https://github.com/tenstorrent/tt-metal/actions/runs/28402940514/job/84173248485)
- **Most recent failing:** commit `25ee250d2211` — 2026-06-29T21:10:57Z — [job](https://github.com/tenstorrent/tt-metal/actions/runs/28402943907/job/84179318567)

### Failing job URLs (last 4 runs)

- https://github.com/tenstorrent/tt-metal/actions/runs/28402943907/job/84179318567
- https://github.com/tenstorrent/tt-metal/actions/runs/28402942896/job/84175236763
- https://github.com/tenstorrent/tt-metal/actions/runs/28402941621/job/84179990988
- https://github.com/tenstorrent/tt-metal/actions/runs/28402940514/job/84173248485

### Failing test path(s)

- `tests/ttnn/unit_tests/operations/eltwise/test_math.py::test_cbrt[w=128-h=64]`

### Error excerpt (terminal failure from logs)

```
FAILED tests/ttnn/unit_tests/operations/eltwise/test_math.py::test_cbrt[w=128-h=64] - AssertionError: Max ULP Delta: 2.0 @ [47, 119] = |0.0625 - 0.06201171875| / 0.000244140625
```

Full traceback from logs:
```
>       assert ulp_passed, ulp_message
E       AssertionError: Max ULP Delta: 2.0 @ [47, 119] = |0.0625 - 0.06201171875| / 0.000244140625

tests/ttnn/utils_for_testing.py:258: AssertionError
```

### Reproduction steps

On a `wh_n300` device with fast runtime mode:

```bash
pytest --timeout 300 tests/ttnn/unit_tests/operations/eltwise/test_math.py::test_cbrt -xv -k "w=128-h=64"
```

Or to reproduce as part of the full group:

```bash
pytest --timeout 300 tests/ttnn/unit_tests/operations/eltwise -xv --splits 4 --group 3 -m "not disable_fast_runtime_mode"
```

### Notes

- The test validates the `ttnn.cbrt` (cube root) operation's numerical accuracy using ULP (Units in the Last Place) comparison.
- The failure occurs at tensor position `[47, 119]`: computed value `0.06201171875` vs expected `0.0625`, giving a ULP delta of `2.0` which exceeds the allowed threshold of `1.0`.
- This is a numerical accuracy regression — the `cbrt` SFPU kernel is producing slightly incorrect results for certain input values.
- The failure is byte-for-byte identical across all 4 runs (same position, same values, same ULP delta), confirming it is deterministic and not hardware noise.

**Determinism check:**
- Run 1 (commit `25ee250d2211`): `FAILED tests/ttnn/unit_tests/operations/eltwise/test_math.py::test_cbrt[w=128-h=64] - AssertionError: Max ULP Delta: 2.0 @ [47, 119] = |0.0625 - 0.06201171875| / 0.000244140625`
- Run 2 (commit `e2f8ee5c15`): `FAILED tests/ttnn/unit_tests/operations/eltwise/test_math.py::test_cbrt[w=128-h=64] - AssertionError: Max ULP Delta: 2.0 @ [47, 119] = |0.0625 - 0.06201171875| / 0.000244140625`
- Run 3 (commit `50cc6f82c6`): `FAILED tests/ttnn/unit_tests/operations/eltwise/test_math.py::test_cbrt[w=128-h=64] - AssertionError: Max ULP Delta: 2.0 @ [47, 119] = |0.0625 - 0.06201171875| / 0.000244140625`
- Run 4 (commit `7f36d0321365`): `FAILED tests/ttnn/unit_tests/operations/eltwise/test_math.py::test_cbrt[w=128-h=64] - AssertionError: Max ULP Delta: 2.0 @ [47, 119] = |0.0625 - 0.06201171875| / 0.000244140625`

<!-- AUTO-TRIAGE-METADATA-START -->
`Auto-triage-workflow: Sanity tests`
`Auto-triage-job-name: ttnn-unit-tests / ttnn eltwise group 3 [wh_n300_civ2]`
<!-- AUTO-TRIAGE-METADATA-END -->

<!-- AUTO-TRIAGE-INFO-START -->
## About this issue (automated CI auto-triage)
This issue was opened and assigned by an automated AI system. It posts through a bot token, so notifications may look like they come from a person, but they are automated.

The owner is selected from recent commit history and job-ownership data. It reflects who is *likely able to fix or route* this, not necessarily who caused it, so it can be imperfect.

**If this was assigned to the wrong person, you can:**
- Comment `REASSIGN OWNER: <why this is outside your area / which team or test-area it belongs to>` and the system will automatically re-route it to a better-suited owner.
- Unassign yourself (this stops notifications to you).
- Close the issue if it is not a real or relevant failure.

The assigned owner is reminded about every 3 days until the issue is resolved. Full guide: https://tenstorrent.atlassian.net/wiki/spaces/MI6/pages/2545221692/Auto-Triage+Info
<!-- AUTO-TRIAGE-INFO-END -->

<!-- PIPELINE-REPORT-START -->
## Pipeline report
Analysis report: https://github.com/tenstorrent/tt-metal/actions/runs/28405702263
Slack notification: https://tenstorrent.slack.com/archives/C0B1F8Z7GQM/p1782771103709279?thread_ts=1782770636.586539&cid=C0B1F8Z7GQM
<!-- PIPELINE-REPORT-END -->

--- a/models/common/utility_functions.py
+++ b/models/common/utility_functions.py
@@ -629,7 +629,11 @@ def comp_ulp(golden, calculated, ulp_threshold, allow_nonfinite=False):
     if not within_threshold:
         ulp_index = torch.argmax(ulp_tensor)
         ulp_index_tuple = tuple(int(idx) for idx in torch.unravel_index(ulp_index, golden.shape))
-        message += f" @ {list(ulp_index_tuple)} = |{calculated[ulp_index_tuple]} - {golden[ulp_index_tuple]}| / {ulp_value[ulp_index_tuple]}"
+        message += (
+            f" @ {list(ulp_index_tuple)} = "
+            f"|calculated {calculated[ulp_index_tuple]} - golden {golden[ulp_index_tuple]}| "
+            f"/ ULP(golden) {ulp_value[ulp_index_tuple]}"
+        )
     return (within_threshold, message)
 
 

--- a/tests/ttnn/unit_tests/operations/eltwise/test_math.py
+++ b/tests/ttnn/unit_tests/operations/eltwise/test_math.py
@@ -159,10 +159,25 @@ def test_rad2deg(device, h, w):
     run_math_unary_test(device, h, w, ttnn.rad2deg, ulp=1)
 
 
-@pytest.mark.parametrize("h", [64])
-@pytest.mark.parametrize("w", [128])
-def test_cbrt(device, h, w):
-    run_math_unary_test(device, h, w, ttnn.cbrt, ulp=1)
+def test_cbrt(device):
+    # Exhaustive over every bf16 value so coverage can't drift with the torch RNG version.
+    # Evaluate the registered golden in fp64 by upcasting the input: it is sgn(x)*pow(|x|,1/3),
+    # and in bf16 the non-representable 1/3 rounds the reference up to 2 ULP short of the true
+    # cube root while the kernel is correct. Subnormal bf16 inputs are flushed to zero on device,
+    # so flush the reference input to match.
+    all_bitpatterns = torch.arange(0, 2**16, dtype=torch.int32).to(torch.uint16)
+    input_tensor = all_bitpatterns.view(torch.bfloat16)
+    input_tensor = input_tensor[torch.isfinite(input_tensor.to(torch.float32))]
+    input_tensor = torch.where(
+        input_tensor.to(torch.float32).abs() < 2.0**-126, torch.zeros_like(input_tensor), input_tensor
+    )
+
+    golden_function = ttnn.get_golden_function(ttnn.cbrt)
+    golden = golden_function(input_tensor.to(torch.float64))
+
+    tt_in = ttnn.from_torch(input_tensor, dtype=ttnn.bfloat16, device=device, layout=ttnn.TILE_LAYOUT)
+    result = ttnn.to_torch(ttnn.cbrt(tt_in))
+    assert_with_ulp(golden, result, 1)
 
 
 @pytest.mark.parametrize("h", [64])

--- a/tests/ttnn/unit_tests/operations/eltwise/test_math.py
+++ b/tests/ttnn/unit_tests/operations/eltwise/test_math.py
@@ -160,11 +160,10 @@ def test_rad2deg(device, h, w):
 
 
 def test_cbrt(device):
-    # Exhaustive over every bf16 value so coverage can't drift with the torch RNG version.
-    # Evaluate the registered golden in fp64 by upcasting the input: it is sgn(x)*pow(|x|,1/3),
-    # and in bf16 the non-representable 1/3 rounds the reference up to 2 ULP short of the true
-    # cube root while the kernel is correct. Subnormal bf16 inputs are flushed to zero on device,
-    # so flush the reference input to match.
+    # Exhaustive over every bf16 value.
+    # Evaluate the registered golden in fp64 by upcasting the input,
+    # in bf16 the non-representable 1/3 was rounding the reference up to 2 ULP short of the true
+    # cube root while the kernel was correct. Subnormal bf16 inputs are flushed to zero on device.
     all_bitpatterns = torch.arange(0, 2**16, dtype=torch.int32).to(torch.uint16)
     input_tensor = all_bitpatterns.view(torch.bfloat16)
     input_tensor = input_tensor[torch.isfinite(input_tensor.to(torch.float32))]

--- a/tests/ttnn/unit_tests/operations/eltwise/test_unary.py
+++ b/tests/ttnn/unit_tests/operations/eltwise/test_unary.py
@@ -1771,33 +1771,6 @@ def test_unary_cbrt_ttnn(input_shapes, torch_dtype, ttnn_dtype, atol, device):
         assert_allclose(ttnn.to_torch(output_tensor), golden_tensor, rtol=1e-05, atol=atol)
 
 
-def test_cbrt_arange(device):
-    # Generate all possible bit patterns for bf16
-    all_bitpatterns = torch.arange(0, 2**16, dtype=torch.int32)
-
-    input_tensor = all_bitpatterns.to(torch.uint16).view(torch.bfloat16)
-    input_tensor = input_tensor.to(torch.float32)
-
-    # Mask subnormals (they get flushed to zero) and NaN (converted to inf for bf16)
-    mask = (((all_bitpatterns >> 7) & 0xFF) == 0) | ((all_bitpatterns & 0x7F) != 0) | torch.isnan(input_tensor)
-    input_tensor[mask] = 1.0
-
-    tt_in = ttnn.from_torch(
-        input_tensor,
-        dtype=ttnn.bfloat16,
-        device=device,
-        layout=ttnn.TILE_LAYOUT,
-        memory_config=ttnn.DRAM_MEMORY_CONFIG,
-    )
-
-    golden_function = ttnn.get_golden_function(ttnn.cbrt)
-    golden = golden_function(input_tensor, device=device)
-
-    tt_result = ttnn.cbrt(tt_in)
-    result = ttnn.to_torch(tt_result)
-    assert_with_ulp(golden, result, 1, allow_nonfinite=True)
-
-
 @pytest.mark.parametrize("ttnn_op", [ttnn.isinf, ttnn.isnan, ttnn.isposinf, ttnn.isneginf, ttnn.isfinite])
 @pytest.mark.parametrize(
     "torch_dtype, ttnn_dtype",

--- a/tests/ttnn/utils_for_testing.py
+++ b/tests/ttnn/utils_for_testing.py
@@ -177,10 +177,15 @@ def assert_with_ulp(
 
     The error is measured using the following formula:
     ``
-        | expected - actual | / ULP(expected)
+        | actual - expected | / ULP(expected)
     ``
 
     Where ULP(expected) returns, for each element, the length of a single Unit of Least Precision (ULP).
+
+    ``expected_result`` is the reference (golden) tensor and ``actual_result`` is the tensor under test.
+    On failure the message reports the worst element as ``|calculated <actual> - golden <expected>| /
+    ULP(golden)``, i.e. the first printed operand is ``actual_result`` and the divisor is the ULP of
+    ``expected_result``.
 
 
     Args:

